Merge `feature/contact-info` branch into `trunk`

ecgan opened the pull request:
### Changes proposed in this Pull Request:

This PR merges the `feature/contact-info` branch into `trunk`. The changes in the branch have been reviewed in other smaller PRs, so we don't really need to review them in detail here.

(Merges changes for https://github.com/woocommerce/google-listings-and-ads/issues/863)

### Changelog entry

> Add - New contact information feature.

mikkamp committed:
Set source when creating issues

mikkamp committed:
Helper function for getting contact info setup url

mikkamp committed:
Add setting for contact info setup

mikkamp committed:
Helper function for when contact info is setup

eason9487 committed:
Add the fourth step page for MC setup

eason9487 committed:
Change values of MC steps to string type and add the fourth step key/value

eason9487 committed:
Change MC setup to have four steps

eason9487 committed:
Add 'libphonenumber-js' to dependencies

eason9487 committed:
Decode HTML entities within countey names

eason9487 committed:
Add a hook to get all supported country calling codes

eason9487 committed:
Add a hook to get user's MC phone number

eason9487 committed:
Add 'hideIcon' prop to AccountCard and make its className extendable

eason9487 committed:
Add new appearance to AccountCard for phone number

eason9487 committed:
Implement the shared components of PhoneNumberCard and ContactInformation section

mikkamp committed:
Notification when contact info is not set

eason9487 committed:
Fix icon size and gap between icon and text when icon on right side for Button

eason9487 committed:
Add a hook to get store address via the core options API

eason9487 committed:
Add new appearance to AccountCard for store address

eason9487 committed:
Implement the shared StoreAddressCard component

eason9487 committed:
Add StoreAddressCard to ContactInformation

mikkamp committed:
Increase action_url size in DB table

mikkamp committed:
Add issue to table when contact info is not set

nima-karimi committed:
Add API endpoints to update and fetch contact information

The previously added API endpoints for updating and fetching the phone number are now modified to also handle updating the business address.

The endpoint name is changed from `phone-number` to `contact-information` but it still only accepts one `phone_number` input. However, now it also updates the address in Merchant Center to the store's address set in WooCommerce settings.

The `GET` endpoint now also returns three extra parameters in its response:
- `mc_address`: The address stored in Merchant Center
- `wc_address`: The address stored in WooCommerce Settings
- `is_mc_address_different`: A boolean indicating whether the two addresses above are different

This also sets the country for the Merchant Center account but given that the MC accounts created by GLA are all sub-accounts and [Google does not allow changing a sub-account's country](https://developers.google.com/shopping-content/reference/rest/v2.1/accounts#AccountAddress) we might need to modify this once the Google API behaviour is changed. Google API currently doesn't enforce this and allows modifying the country (even though it's not possible to do so via MC interface).

Merchant Center API only accepts one address line and to adapt the WooCommerce address to it, the two addresses entered in WooCommerce settings are merged into one string separated with a new line character (`\n`). MC interface detects the addresses correctly and displays each in their own fields.

Note that address components are returned in the API response as they appear on Google API. This means that their naming might be different from WooCommerce. For example, city is called `locality`, state is called `region`, etc. This is just for the sake of simplicity and can be modified if we want to keep things similar to WC core.

mikkamp committed:
Check MC to determine if contact info is setup

nima-karimi committed:
Move address comparison to a utility class

nima-karimi committed:
Make sure the addresses aren't null before comparison

eason9487 committed:
Add URLs to utils for settings and edit contact info pages

nima-karimi committed:
Add unit tests for AddressUtility class

eason9487 committed:
Implement ContactInformationPreview for the main page of settings

eason9487 committed:
Implement the EditContactInformation page

eason9487 committed:
Add ContactInformationPreview and EditContactInformation to settings page

eason9487 committed:
Update wording

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#issuecomment-889723783

eason9487 committed:
Update the URL of "Learn more" links

mikkamp committed:
Check address and phone number

mikkamp committed:
Clear transient when updating contact details

nima-karimi committed:
Add missing dependency for unit test

nima-karimi committed:
Create separate methods to update phone number and address

Only update the phone number if a value is set in the request params. If a null or empty value is set then remove the phone number from contact information.

Rename the `MerchantVerification` service to `ContactInformation`.

mikkamp committed:
Add transients to MerchantVerification test

mikkamp committed:
Rename MerchantVerification

mikkamp committed:
Save option once user has been fully setup

nima-karimi committed:
Update tests/Unit/API/Site/Controllers/MerchantCenter/ContactInformationControllerTest.php

Co-authored-by: Mik <mikkamp@users.noreply.github.com>

nima-karimi committed:
Merge pull request #919 from woocommerce/feature/contact-info-rest-api

Add API endpoints to update and fetch contact information

eason9487 committed:
Merge pull request #914 from woocommerce/feature/863-mc-setup-flow

Change MC setup flow to have four steps

mikkamp committed:
Update notification wording

eason9487 committed:
Integrate data store with the backend contact info API

eason9487 committed:
Change the data source of useStoreAddress to the contact info API

eason9487 committed:
Add isValid and isDirty to the callback param of onPhoneNumberChange

eason9487 committed:
Implement the data submission of EditContactInformation

eason9487 committed:
Implement the data submission of StoreRequirements

mikkamp committed:
Revert transient and rely only on option

mikkamp committed:
Revert unnecessary changes in test

eason9487 committed:
Move PreLaunchChecklist and related code out from the step 3 of MC setup

eason9487 committed:
Add PreLaunchChecklist to the step 4 of MC setup

mikkamp committed:
Merge pull request #912 from woocommerce/feature/910-contact-info-notification

Contact info notification and issue

layoutd committed:
Add fourth value for MC Setup step 'store_requirements'

eason9487 committed:
Rename parameter and simplify conditional `setState` calls

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681013079

eason9487 committed:
Use @wordpress/html-entities instead

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681036085
- https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681038414

eason9487 committed:
Rename useCountryCallingCodes to useCountryCallingCodeOptions and remove `as` parameter

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681073767

eason9487 committed:
Remove useGoogleMCPhoneNumber from PhoneNumberCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681154403

eason9487 committed:
Rename className for AppButton

Address https://github.com/woocommerce/google-listings-and-ads/pull/918#discussion_r681169915

nima-karimi committed:
Move the store address check

eason9487 committed:
Hold rendering before settings data loaded for StoreRequirements

eason9487 committed:
Split tests of unset value and invalid value.

Address https://github.com/woocommerce/google-listings-and-ads/pull/927#discussion_r681095105

eason9487 committed:
Remove the TODO comment for 'store_requirements' constant due to backend took the same one.

Refer to https://github.com/woocommerce/google-listings-and-ads/pull/928

tomalec committed:
Refactor `…/pre-launch-checklist/checkErrors.test.js `.

Make descriptions more descriptive, remove unused parameters,
remove snapshots.

eason9487 committed:
Make the icon a horizontal center of focus area

Address https://github.com/woocommerce/google-listings-and-ads/pull/918#discussion_r681924240

Co-authored-by: Tomek Wytrębowicz <tomalecpub@gmail.com>

eason9487 committed:
Use SpinnerCard instead

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#discussion_r681954844

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Use SpinnerCard instead for the DisconnectAccounts to make all loading UIs consistency in the Setting page

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#discussion_r681954844

eason9487 committed:
Simplify the else condition for ContactInformationPreview

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#discussion_r681952956

eason9487 committed:
Add event tracking for the "Learn more" link in the NoContactInformationCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#discussion_r681979584

eason9487 committed:
Merge pull request #917 from woocommerce/feature/863-mc-phone-number

Implement `PhoneNumberCard` and `ContactInformation` components for contact info

eason9487 committed:
Merge pull request #918 from woocommerce/feature/863-mc-store-address

Implement `StoreAddressCard` component for contact info

eason9487 committed:
Pass phone number data as phoneNumber prop to PhoneNumberCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/921#discussion_r681414104

eason9487 committed:
Merge pull request #921 from woocommerce/feature/863-settings-contact-info

Add contact information section and its edit subpage to the settings page

eason9487 committed:
Merge pull request #926 from woocommerce/feature/863-integrate-api

Integrate with contact information API and complete form submissions within MC setup and settings

eason9487 committed:
Merge pull request #927 from woocommerce/feature/863-pre-launch-checklist

Move the pre-launch checklist from step 3 to 4 for the MC setup

eason9487 committed:
Handle loading state within PhoneNumberCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681624815

eason9487 committed:
Add JSDoc for PhoneNumberCard

layoutd committed:
Don't default to `flat` for shipping settings check

layoutd committed:
Comment fixes

layoutd committed:
!isset instead of is_null

layoutd committed:
Merge pull request #928 from woocommerce/add/four-onboarding-steps

Add fourth MC Setup step value: 'store_requirements'

eason9487 committed:
Show the divider when loading

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#issuecomment-893476476

eason9487 committed:
Add a comment for the handling of initEditing = null

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#discussion_r683568574

eason9487 committed:
Simplify the if else statements

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#discussion_r683571373

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Merge pull request #934 from woocommerce/feature/phone-number-card-loading

Handle all loading states within the `PhoneNumberCard` component

eason9487 committed:
Fix calling .split() on null value in the useStoreAddress

eason9487 committed:
Fix that unable to save if only changes address

eason9487 committed:
Check if the GMC address is filled in the ContactInformationPreview

ecgan committed:
Merge pull request #936 from woocommerce/fix/null-store-address

Fix a bug that calling `split()` on `null` value in the useStoreAddress hook

ecgan committed:
Merge pull request #937 from woocommerce/fix/edit-contact-info-save-logic

tomalec committed:
Make `useStoreAddress` cater for `mc_address` being `null`

as well.

Otherwise, for merchants that didn't set the store address at all, Settings page would throw and error and render empty.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/938#issuecomment-895196693

tomalec committed:
Merge pull request #938 from woocommerce/fix/check-required-mc-address

tomalec commented:
> The changes in the branch have been reviewed in other smaller PRs, so we don't really need to review them in detail here.

But none of CI checks were done (see #922), so it would be nice to check it here.

tomalec commented:
I'm still stuck with an account with an incomplete MC address. It does not allow me to save address, as it "Add contact information" uses WC address, but Settings check MC address:
![wrong address checked](https://user-images.githubusercontent.com/17435/128726108-5c6e0265-cafa-4d04-b1b9-054e1f9424b9.gif)

edit:

I can no longer reproduce that, maybe I had some merge artifacts :\

ecgan committed:
Check whether phone number exists and call track event.

ecgan committed:
Track event when users click on phone number edit button.

tomalec committed:
Track "edit store address button clicked" event,

Implements part of https://github.com/woocommerce/google-listings-and-ads/issues/863.

ecgan committed:
Code refactor with usePhoneNumberCheckTrackEventEffect.

ecgan committed:
Specify view="settings" for PhoneNumberCard in ContactInformationPreview.

ecgan committed:
Add docs for phone number track events.

ecgan committed:
Add docs for edit_mc_store_address track event.

ecgan committed:
Fix docs for edit_mc_store_address track event.

tomalec committed:
Merge pull request #944 from woocommerce/add/863-track-store

ecgan committed:
Use path instead of view in usePhoneNumberCheckTrackEventEffect.

ecgan committed:
Add track event for contact_information_save_button_click.

ecgan committed:
Merge pull request #945 from woocommerce/feature/phone-number-track-event

tomalec committed:
Merge pull request #946 from woocommerce/feature/save-contact-info-track-event
